File: ts-langchain/tests/unit/test_react.ts
import { describe, expect, it } from "vitest";
import { AIMessage } from "@langchain/core/messages";
import { DynamicStructuredTool } from "langchain/tools";
import { z } from "zod";
import {
  createDefaultReActLimits,
  ReActAgentExecutor,
} from "../../src/agents/react.js";

class FakeModel {
  private readonly responses: AIMessage[];
  calls = 0;

  constructor(responses: AIMessage[]) {
    this.responses = [...responses];
  }

  bindTools(): this {
    return this;
  }

  async invoke(): Promise<AIMessage> {
    this.calls += 1;
    const response = this.responses.shift();
    if (!response) throw new Error("no fake response");
    return response;
  }
}

function createWeatherTool(shouldFail = false): DynamicStructuredTool {
  return new DynamicStructuredTool({
    name: "weather.current",
    description: "查询天气",
    schema: z.object({ city: z.string() }),
    func: async () => {
      if (shouldFail) throw new Error("weather unavailable");
      return JSON.stringify({ city: "Shanghai", temperature: 25 });
    },
  });
}

describe("ReAct executor", () => {
  it("reasons again after an Observation and returns a structured summary", async () => {
    const model = new FakeModel([
      new AIMessage({
        content: "",
        tool_calls: [
          { id: "call-weather", name: "weather.current", args: { city: "Shanghai" } },
        ],
      }),
      new AIMessage("上海现在 25°C。"),
    ]);
    const executor = new ReActAgentExecutor(
      model,
      [createWeatherTool()],
      "test prompt",
      createDefaultReActLimits({ maxTotalTimeMs: 10_000 }),
    );

    const result = await executor.invoke({ input: "上海天气怎么样？" });

    expect(result.output).toContain("25°C");
    expect(result.react.stop_reason).toBe("ANSWER_COMPLETE");
    expect(result.react.tool_calls).toBe(1);
    expect(result.react.observations[0].status).toBe("success");
  });

  it("retries the same failed call once and then stops", async () => {
    let callNumber = 0;
    const call = () =>
      new AIMessage({
        content: "",
        tool_calls: [
          { id: `call-weather-${++callNumber}`, name: "weather.current", args: { city: "Shanghai" } },
        ],
      });
    const model = new FakeModel([call(), call(), call()]);
    const executor = new ReActAgentExecutor(
      model,
      [createWeatherTool(true)],
      "test prompt",
      createDefaultReActLimits({ maxTotalTimeMs: 10_000 }),
    );

    const result = await executor.invoke({ input: "上海天气怎么样？" });

    expect(result.react.stop_reason).toBe("TOOL_FAILURE");
    expect(result.react.tool_calls).toBe(2);
    expect(result.react.tool_errors).toBe(2);
  });
});

File: py-langgraph/tests/agents/test_react.py
from langchain_core.messages import ToolMessage

from src.agents.react import (
    is_clarification,
    observation_from_tool_message,
    summarize_react_state,
    tool_call_signature,
)


def test_react_signature_is_stable_for_argument_order():
    assert tool_call_signature("web.search", {"b": 2, "a": 1}) == tool_call_signature(
        "web.search", {"a": 1, "b": 2}
    )


def test_tool_message_becomes_observation_envelope():
    message = ToolMessage(
        content='{"status":"success","data":{"value":4},"error":null}',
        tool_call_id="call-1",
        name="calculator",
    )

    observation = observation_from_tool_message(message)

    assert observation.tool_call_id == "call-1"
    assert observation.tool == "calculator"
    assert observation.status == "success"
    assert observation.data == {"value": 4}


def test_react_summary_and_clarification_detection():
    assert is_clarification("请问你想查询哪个城市的天气？")
    summary = summarize_react_state(
        {
            "react_state": "COMPLETED",
            "stop_reason": "ANSWER_COMPLETE",
            "react_steps": 2,
            "react_tool_calls": 1,
            "react_tool_names": ["weather.current"],
            "tool_errors": 0,
            "model_calls": 2,
            "observations": [],
            "total_latency_ms": 12,
        }
    )
    assert summary["stop_reason"] == "ANSWER_COMPLETE"
    assert summary["tool_calls"] == 1
